rpcs3/Emu/System.cpp

```diff
@@ -1657,25 +1657,65 @@ void Emulator::Resume()
 	}
 }
 
-void Emulator::Stop(bool restart)
+u32 sysutil_send_system_cmd(u64 status, u64 param);
+void process_qt_events();
+
+void Emulator::GracefulShutdown(bool allow_autoexit, bool async_op)
 {
-	if (m_state.exchange(system_state::stopped) == system_state::stopped)
+	const auto old_state = m_state.load();
+
+	if (old_state == system_state::stopped)
+	{
+		return;
+	}
+
+	if (old_state == system_state::paused)
+	{
+		Resume();
+	}
+
+	if (old_state == system_state::frozen || !sysutil_send_system_cmd(0x0101 /* CELL_SYSUTIL_REQUEST_EXITGAME */, 0))
+	{
+		// The callback has been rudely ignored, we have no other option but to force termination
+		Kill(allow_autoexit);
+		return;
+	}
+
+	auto perform_kill = [allow_autoexit, this, info = ProcureCurrentEmulationCourseInformation()]()
 	{
-		if (restart)
+		for (u32 i = 0; i < 50; i++)
```

Comment:
lmao, just realized this is blocking for 5 seconds before even going for the kill.
It's also missing an exit condition

Reply:
Fixed.